cupy/cuda/cupy_thrust.cu

```diff
@@ -43,6 +44,102 @@ public:
 };
 
 
+/* ------------------------------------ Minimum boilerplate to support complex numbers ------------------------------------ */
+// We need a specialized operator< here in order to match the NumPy behavior:
+// "The sort order for complex numbers is lexicographic. If both the real and imaginary parts are non-nan then the order is
+// determined by the real parts except when they are equal, in which case the order is determined by the imaginary parts.
+//
+// In numpy versions >= 1.4.0 nan values are sorted to the end. The extended sort order is:
+//     Real: [R, nan]
+//     Complex: [R + Rj, R + nanj, nan + Rj, nan + nanj]
+// where R is a non-nan real value. Complex values with the same nan placements are sorted according to the non-nan part if
+// it exists. Non-nan values are sorted as before."
+// Ref: https://docs.scipy.org/doc/numpy/reference/generated/numpy.sort.html
+
+__host__ __device__ inline bool isnan(const cuComplex& z) {
+    return isnan(z.x) || isnan(z.y);
+}
+
+__host__ __device__ inline bool isnan(const cuDoubleComplex& z) {
+    return isnan(z.x) || isnan(z.y);
+}
+
+__host__ __device__ inline bool operator<(const cuComplex& lhs,
+                                          const cuComplex& rhs) {
```

Comment:
Probably is good to cache the result of the isnan calls in bools instead of calling them multiple times for a comparisson. After all it ends relying on a external "isnan" function that I am not sure if the compile can cache and reuse the result multiple times.

Reply:
For floating-point numbers, `isnan` on host is from standard C99 library, and on device is from CUDA math library, so it's standard everywhere. For most sane implementations, It is a very cheap operation (only one load + one comparison):
```C
bool isnan(float x) {
    return x != x;
}
```
So it's not worth extra effort, as caching it still takes one load + one comparison (for `if`):
```C
bool ReZ_nan = isnan(z.x);
if (ReZ_nan) {
    // do something
}
```

Reply:
I ended up caching it in 668efd0...😞